File: code_reality/graph_audit.py
"""CRG graph.db Rust 完整度稽核——D1 風險掃描＋D2 rust-analyzer 對帳。

收編自 NT N1 repo-local 腳本（2026-08-24 驗證輪：239 風險檔／187 對帳／
219 確定缺差——kernel.rs 九項全中；同日 NT 三輪獨立審查迭代版同步：縮排
impl 可見、閉合縮排對齊、RA 失敗不崩潰＋聚合假乾淨防護）。偵測 CRG
``nodes.qualified_name UNIQUE``＋``ON CONFLICT DO UPDATE`` 後寫蓋前寫的
靜默丟失：Rust 型別的 inherent impl 與 trait impl 同名方法同鍵 ``X.method``，
前者消失（``head_matches_build`` 新鮮度指標不反映此項——檔案未變不重
parse；根因記錄見 NT memory crg-graph-build-node-loss）。

兩層偵測：
  D1 風險掃描（純文字，秒級）：同名方法出現於 ≥2 個 impl 塊的檔案。
    **per-block 計數非全體交集**——交集會被 Drop 等單方法 impl 清空
    （kernel.rs 三 impl〔inherent＋Drop＋trait〕漏報實證）。縮排 impl
    （inline ``mod`` 內）可匹配；impl 閉合＝與 impl 關鍵字同縮排的 ``}``，
    無法閉合時方法歸屬保守膨脹
  D2 對帳（每檔百毫秒級）：rust-analyzer ``symbols`` stdin 模式（獨立源）
    vs graph.db nodes 每名計數——DB 少於 rust-analyzer＝被去重吃掉。
    **DB 側 kind 須含 'Test'**——CRG 把 #[test] 函數建為 Test 節點、
    rust-analyzer 側是 Function/Method，漏計＝全部測試函數誤報缺差
    （NT N1 首跑 1,670 假警報實證）

已知保守偏差（D1 只膨脹風險清單、不漏）：blanket impl（``impl<T> Trait
for T``）與 ``impl Trait for Vec<X>`` 歸到 ``T``/``Vec`` 名下；``dyn`` 型別
已處理、泛型路徑追蹤從簡。

用法::

    uv run --project ~/Github/ai-rules python -m code_reality.graph_audit \
        --repo <repo> [--all] [--json] [--graph PATH]

掃描集＝profile ``[[scan_root]]`` 的 path glob（Rust 形態 repo）；無
scan_root → repo 全 ``*.rs`` 經 exclusions 過濾（generic fallback）。
graph 預設 ``<repo>/.code-review-graph/graph.db``。退出碼：0=乾淨｜
1=發現缺差｜2=環境錯誤（rust-analyzer 未裝／graph.db 不在／全部檔案 RA
解析失敗——假乾淨防護）。``--json`` 鍵（risk_files/audited_files/
missing/errors）為 NT 治理鉤子契約（errors 鍵＝NT 最終版新增）。
"""

import argparse
import json
import re
import shutil
import subprocess
import sys
from collections import Counter
from pathlib import Path

from code_reality.common import connect_ro, graph_db_path
from code_reality.exclusions import is_excluded
from code_reality.profile import load_profile, scan_roots

# D1 正則（NT 三輪獨立審查迭代後形態）：前綴容許 unsafe；泛型段 [^{]* 吞
# 巢狀 >；trait 容許路徑限定 fmt::Display；for 型容許 dyn Foo
IMPL_RE = re.compile(
    r"^\s*(?:unsafe\s+)?impl(?:<[^{]*>)?\s+"
    r"(?:(?:\w+::)*\w+\s+for\s+)?((?:dyn\s+)?[A-Z]\w*)"
)
FN_RE = re.compile(
    r"^\s*(?:pub(?:\([^)]*\))?\s+)?"
    r"(?:const\s+|async\s+|unsafe\s+|extern(?:\s*\"[^\"]*\")?\s+)*fn\s+(\w+)"
)
RA_LABEL_RE = re.compile(r'label: "([^"]*)"')
RA_KIND_RE = re.compile(r"kind: SymbolKind\((\w+)\)")
RA_FN_KINDS = {"Function", "Method"}


def scan_files(repo: Path) -> list[Path]:
    """掃描集：profile [[scan_root]] path glob；無 → repo 全 *.rs 經
    exclusions 過濾（generic fallback——與 PathResolver/pkg_roots 同型）。"""
    profile = load_profile(repo)
    roots = scan_roots(profile)
    if roots:
        return sorted({p for sr in roots for p in repo.glob(sr.path)})
    return sorted(
        p
        for p in repo.rglob("*.rs")
        if not is_excluded(p.relative_to(repo).as_posix(), profile)
    )


def risk_scan(files: list[Path]) -> list[tuple[Path, str, list[str]]]:
    """D1：同名方法出現於 ≥2 個 impl 塊的檔案（不碰 graph）。

    準則＝任兩塊碰撞（每方法計「出現於幾個 impl 塊」，≥2 即候選）——非全體
    交集（會被 Drop 等單方法 impl 清空——kernel.rs 三 impl 實證漏報）。
    impl 閉合偵測＝與 impl 關鍵字同縮排的 ``}``；無法閉合時方法歸屬保守膨脹。
    """
    at_risk: list[tuple[Path, str, list[str]]] = []
    for f in files:
        impls: list[list] = []  # [type, [fn names], indent]
        cur: list | None = None
        for line in f.read_text(encoding="utf-8", errors="replace").splitlines():
            m = IMPL_RE.match(line)
            if m:
                cur = [m.group(1), [], len(line) - len(line.lstrip())]
                impls.append(cur)
                continue
            if cur is None:
                continue
            stripped = line.strip()
            if stripped == "}":
                if len(line) - len(line.lstrip()) <= cur[2]:
                    cur = None
                continue
            fm = FN_RE.match(line)
            if fm:
                cur[1].append(fm.group(1))
        block_counts: dict[str, Counter] = {}
        for entry in impls:
            counts = block_counts.setdefault(entry[0], Counter())
            for name in set(entry[1]):
                counts[name] += 1
        for t, counts in block_counts.items():
            overlap = sorted(n for n, c in counts.items() if c >= 2)
            if overlap:
                at_risk.append((f, t, overlap))
    return at_risk


def parse_ra_symbols(stdout_text: str) -> Counter:
    """rust-analyzer ``symbols`` 輸出 → (label → fn 計數)；非 fn kind 略過。"""
    counts: Counter = Counter()
    for line in stdout_text.splitlines():
        kind = RA_KIND_RE.search(line)
        if not kind or kind.group(1) not in RA_FN_KINDS:
            continue
        label = RA_LABEL_RE.search(line)
        if label:
            counts[label.group(1)] += 1
    return counts


def ra_symbols(path: Path) -> Counter | None:
    """rust-analyzer stdin 模式（獨立於 CRG parser 的對帳源）。

    逾時/失敗回 None（caller 記 errors 不崩潰）；``check=False``＝非零退出
    輸出空清單——vacuous pass（零比較）非全缺差，audit 迴圈對非空檔零輸出
    印 [WARN]、main 聚合零符號擋 exit 2（雙層防靜默假陰性）。
    """
    try:
        proc = subprocess.run(
            ["rust-analyzer", "symbols"],
            input=path.read_bytes(),
            capture_output=True,
            timeout=60,
            check=False,
        )
    except subprocess.TimeoutExpired:
        return None
    return parse_ra_symbols(proc.stdout.decode("utf-8", errors="replace"))


def db_functions(conn, path: Path) -> Counter:
    """graph.db nodes 每名計數——kind 含 'Test'（漏計＝測試函數全誤報）。"""
    rows = conn.execute(
        "SELECT name, COUNT(*) FROM nodes WHERE file_path = ? "
        "AND kind IN ('Function', 'Test') GROUP BY name",
        (str(path.resolve()),),
    )
    return Counter({name: count for name, count in rows})


def audit(
    repo: Path,
    graph: Path,
    *,
    all_files: bool = False,
    ra_lookup=None,
) -> tuple[
    list[tuple[Path, str, list[str]]], int, list[dict[str, object]], list[str], int
]:
    """D1＋D2 主流程——回 (risk, audited_count, missing, errors, total_ra)。

    ``ra_lookup`` 注入點供測試替身（預設 rust-analyzer subprocess；回 None
    ＝逾時/失敗記 errors 跳過）。
    """
    files = scan_files(repo)
    risk = risk_scan(files)
    scope = files if all_files else sorted({f for f, _, _ in risk})
    lookup = ra_lookup or ra_symbols
    conn = connect_ro(graph)
    try:
        missing: list[dict[str, object]] = []
        errors: list[str] = []
        total_ra = 0
        for f in scope:
            ra = lookup(f)
            if ra is None:
                errors.append(f"{f}: rust-analyzer 逾時/失敗（跳過）")
                continue
            total_ra += sum(ra.values())
            db = db_functions(conn, f)
            if not ra and f.stat().st_size > 0:
                print(
                    f"[WARN] rust-analyzer 對 {f.name} 零輸出（格式 drift 或"
                    "單檔 parse fail）——該檔對帳 vacuous，勿當乾淨讀",
                    file=sys.stderr,
                )
            for name, ra_count in ra.items():
                db_count = db.get(name, 0)
                if db_count < ra_count:
                    missing.append(
                        {
                            "file": str(f),
                            "symbol": name,
                            "ra_count": ra_count,
                            "db_count": db_count,
                        }
                    )
    finally:
        conn.close()
    return risk, len(scope), missing, errors, total_ra


def main() -> int:
    parser = argparse.ArgumentParser(description="CRG graph.db Rust 完整度稽核")
    parser.add_argument("--repo", type=Path, required=True, help="掃描目標 repo 根")
    parser.add_argument(
        "--all", action="store_true", help="對帳全部 .rs（預設僅風險檔）"
    )
    parser.add_argument(
        "--json", action="store_true", help="機器可讀輸出（治理鉤子契約）"
    )
    parser.add_argument(
        "--graph",
        type=Path,
        default=None,
        help="覆寫 graph.db 路徑（預設 <repo>/.code-review-graph/graph.db）",
    )
    args = parser.parse_args()

    if shutil.which("rust-analyzer") is None:
        print(
            "[FAIL] rust-analyzer 不在 PATH——rustup component add rust-analyzer",
            file=sys.stderr,
        )
        return 2
    graph = args.graph if args.graph is not None else graph_db_path(args.repo)
    if not graph.exists():
        print(
            f"[FAIL] graph.db 不存在：{graph}（完整度稽核需要它；"
            "新鮮度指標不保證存在）",
            file=sys.stderr,
        )
        return 2

    risk, audited, missing, errors, total_ra = audit(
        args.repo, graph, all_files=args.all
    )
    # 假乾淨防護：RA 輸出格式漂移或環境異常時 total 為 0——回 2 而非 0
    if audited and total_ra == 0:
        print(
            "[FAIL] 全部檔案 rust-analyzer 符號數為 0——輸出格式漂移或環境錯誤",
            file=sys.stderr,
        )
        return 2

    if args.json:
        print(
            json.dumps(
                {
                    "risk_files": [
                        {"file": str(f), "type": t, "overlap": o} for f, t, o in risk
                    ],
                    "audited_files": audited,
                    "missing": missing,
                    "errors": errors,
                },
                ensure_ascii=False,
                indent=1,
            )
        )
    else:
        print(f"[OK] D1 風險掃描：{len(risk)} 檔（同名方法 ≥2 impl 塊）")
        print(
            f"[OK] D2 對帳：{audited} 檔（rust-analyzer vs graph.db，{total_ra} 符號）"
        )
        for e in errors:
            print(f"[WARN] {e}")
        if missing:
            print(
                f"[WARN] DB 缺差 {len(missing)} 項（同鍵去重吃掉——"
                "head_matches_build 不反映此項）："
            )
            by_file: dict[str, list[dict[str, object]]] = {}
            for m in missing:
                by_file.setdefault(str(m["file"]), []).append(m)
            for f, items in sorted(by_file.items()):
                syms = ", ".join(
                    f"{m['symbol']}({m['db_count']}/{m['ra_count']})" for m in items
                )
                print(f"  {f}: {syms}")
        else:
            print("[OK] 無缺差")

    return 1 if missing else 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/code-reality/src/graph_audit.rs
//! `graph_audit` — the frozen `code_reality/graph_audit.py` contract:
//! D1 risk scan (same-name methods across ≥2 impl blocks) + D2
//! rust-analyzer reconciliation against graph.db nodes. Exit family
//! 0/1/2 (D3): 0 clean | 1 missing found | 2 environment error
//! (rust-analyzer absent / db missing / vacuous all-zero guard).
//! `--json` four keys (risk_files/audited_files/missing/errors) are the
//! governance-hook contract face — stdout bytes gated.

use crate::argparse::{parse, required, FlagSpec, Kind, Outcome, ToolSpec};
use crate::common::{connect_ro, graph_db_path};
use crate::profile::{is_excluded, load_profile, scan_roots};
use crate::ToolOutput;
use rusqlite::Connection;
use std::collections::HashMap;
use std::io::Read;
use std::path::{Path, PathBuf};

const SPEC: ToolSpec = ToolSpec {
    flags: &[
        FlagSpec {
            long: "--repo",
            short: None,
            kind: Kind::Value { metavar: "REPO" },
        },
        FlagSpec {
            long: "--all",
            short: None,
            kind: Kind::StoreTrue,
        },
        FlagSpec {
            long: "--json",
            short: None,
            kind: Kind::StoreTrue,
        },
        FlagSpec {
            long: "--graph",
            short: None,
            kind: Kind::Value { metavar: "GRAPH" },
        },
    ],
    positionals: &[],
};

const HELP: &str = concat!(
    "usage: graph_audit [-h] --repo REPO [--all] [--json] [--graph GRAPH]\n",
    "\n",
    "CRG graph.db Rust 完整度稽核\n",
    "\n",
    "options:\n",
    "  -h, --help     show this help message and exit\n",
    "  --repo REPO    掃描目標 repo 根\n",
    "  --all          對帳全部 .rs（預設僅風險檔）\n",
    "  --json         機器可讀輸出（治理鉤子契約）\n",
    "  --graph GRAPH  覆寫 graph.db 路徑（預設 <repo>/.code-review-graph/graph.db）\n",
);

fn impl_re() -> &'static regex::Regex {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| {
        regex::Regex::new(
            r"^\s*(?:unsafe\s+)?impl(?:<[^{]*>)?\s+(?:(?:\w+::)*\w+\s+for\s+)?((?:dyn\s+)?[A-Z]\w*)",
        )
        .unwrap()
    })
}

fn fn_re() -> &'static regex::Regex {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| {
        regex::Regex::new(
            r#"^\s*(?:pub(?:\([^)]*\))?\s+)?(?:const\s+|async\s+|unsafe\s+|extern(?:\s*"[^"]*")?\s+)*fn\s+(\w+)"#,
        )
        .unwrap()
    })
}

fn ra_kind_re() -> &'static regex::Regex {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| regex::Regex::new(r#"kind: SymbolKind\((\w+)\)"#).unwrap())
}

fn ra_label_re() -> &'static regex::Regex {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| regex::Regex::new(r#"label: "([^"]*)""#).unwrap())
}

const RA_FN_KINDS: [&str; 2] = ["Function", "Method"];

/// Insertion-ordered counter (Python `Counter` iteration semantics):
/// the `--json` `missing` array order = scope file order × first-seen
/// label order within each file (D4).
#[derive(Default)]
pub struct OrderedCounter {
    order: Vec<String>,
    counts: HashMap<String, usize>,
}

impl OrderedCounter {
    pub fn bump(&mut self, key: &str) {
        if !self.counts.contains_key(key) {
            self.order.push(key.to_string());
        }
        *self.counts.entry(key.to_string()).or_insert(0) += 1;
    }

    pub fn iter(&self) -> impl Iterator<Item = (&str, usize)> {
        self.order
            .iter()
            .map(|k| (k.as_str(), self.counts[k]))
    }

    pub fn total(&self) -> usize {
        self.counts.values().sum()
    }

    pub fn is_empty(&self) -> bool {
        self.order.is_empty()
    }
}

fn indent_of(line: &str) -> usize {
    line.len() - line.trim_start().len()
}

/// Scan set (`graph_audit.py:68-79`): profile `[[scan_root]]` path globs;
/// without scan roots the whole repo `*.rs` via exclusions. pathlib glob
/// matches dotfiles — glob crate default `require_literal_leading_dot =
/// false` aligns (sorted + dedup as in Python's set collection).
/// Profile load errors propagate (crash-only — Python `graph_audit.py:71`
/// calls `load_profile` bare; swallowing would silently fall back to the
/// generic whole-repo scan on a broken TOML).
pub fn scan_files(repo: &Path) -> Result<Vec<PathBuf>, String> {
    let profile = load_profile(repo)?;
    let roots = scan_roots(profile.as_ref());
    let mut out: Vec<PathBuf> = if !roots.is_empty() {
        let mut opts = glob::MatchOptions::new();
        opts.require_literal_leading_dot = false;
        roots
            .iter()
            .filter_map(|sr| {
                glob::glob_with(repo.join(&sr.path).to_string_lossy().as_ref(), opts).ok()
            })
            .flatten()
            .filter_map(|r| r.ok())
            .collect()
    } else {
        let mut opts = glob::MatchOptions::new();
        opts.require_literal_leading_dot = false;
        let pattern = repo.join("**/*.rs").to_string_lossy().into_owned();
        glob::glob_with(&pattern, opts)
            .into_iter()
            .flatten()
            .filter_map(|r| r.ok())
            .filter(|p| {
                p.strip_prefix(repo)
                    .map(|rel| !is_excluded(&rel.to_string_lossy(), profile.as_ref()))
                    .unwrap_or(true)
            })
            .collect()
    };
    out.sort();
    out.dedup();
    Ok(out)
}

/// D1 risk scan (`graph_audit.py:82-118`): per-block counting (NOT the
/// global intersection — single-method impls like Drop would empty it);
/// impl closure = `}` at indent ≤ the impl keyword's.
pub fn risk_scan(files: &[PathBuf]) -> Vec<(PathBuf, String, Vec<String>)> {
    let mut at_risk = Vec::new();
    for f in files {
        let text = std::fs::read(f)
            .map(|b| String::from_utf8_lossy(&b).into_owned())
            .unwrap_or_default();
        // impls: (type, [fn names], indent)
        let mut impls: Vec<(String, Vec<String>, usize)> = Vec::new();
        let mut cur: Option<usize> = None; // index into impls
        for line in text.split('\n') {
            if let Some(m) = impl_re().captures(line) {
                impls.push((
                    m[1].to_string(),
                    Vec::new(),
                    indent_of(line),
                ));
                cur = Some(impls.len() - 1);
                continue;
            }
            let Some(idx) = cur else { continue };
            let stripped = line.trim();
            if stripped == "}" {
                if indent_of(line) <= impls[idx].2 {
                    cur = None;
                }
                continue;
            }
            if let Some(fm) = fn_re().captures(line) {
                impls[idx].1.push(fm[1].to_string());
            }
        }
        // per-type aggregated block counts, insertion order preserved
        let mut block_counts: Vec<(String, OrderedCounter)> = Vec::new();
        for (t, names, _) in &impls {
            let entry = match block_counts.iter_mut().find(|(et, _)| et == t) {
                Some((_, c)) => c,
                None => {
                    block_counts.push((t.clone(), OrderedCounter::default()));
                    &mut block_counts.last_mut().unwrap().1
                }
            };
            let mut seen: std::collections::BTreeSet<&str> = Default::default();
            for n in names {
                if seen.insert(n) {
                    entry.bump(n);
                }
            }
        }
        for (t, counts) in &block_counts {
            let mut overlap: Vec<String> = counts
                .iter()
                .filter(|&(_, c)| c >= 2)
                .map(|(n, _)| n.to_string())
                .collect();
            overlap.sort(); // graph_audit.py:115 — sorted() on the overlap face
            if !overlap.is_empty() {
                at_risk.push((f.clone(), t.clone(), overlap));
            }
        }
    }
    at_risk
}

/// rust-analyzer `symbols` stdout → label counts (fn kinds only).
pub fn parse_ra_symbols(stdout_text: &str) -> OrderedCounter {
    let mut counts = OrderedCounter::default();
    for line in stdout_text.split('\n') {
        let Some(kind) = ra_kind_re().captures(line) else {
            continue;
        };
        if !RA_FN_KINDS.contains(&&kind[1]) {
            continue;
        }
        if let Some(label) = ra_label_re().captures(line) {
            counts.bump(&label[1]);
        }
    }
    counts
}

/// rust-analyzer stdin mode (`graph_audit.py:134-151`): the file is passed
/// as stdin (an opened fd — no writer side, so no pipe deadlock), stdout
/// parsed lossily, `check=False`. Timeout (60s) → `Ok(None)` (caller
/// records an error); spawn failure is the crash family (Python would
/// raise FileNotFoundError after the env gate).
pub fn ra_symbols(path: &Path) -> Result<Option<OrderedCounter>, String> {
    let file = std::fs::File::open(path)
        .map_err(|e| format!("{} 開啟失敗：{}", path.display(), e))?;
    let mut child = std::process::Command::new("rust-analyzer")
        .arg("symbols")
        .stdin(std::process::Stdio::from(file))
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::null())
        .spawn()
        .map_err(|e| format!("rust-analyzer spawn 失敗：{}", e))?;
    // drain stdout concurrently so a large symbol list can't fill the pipe
    let mut out_pipe = child.stdout.take().unwrap();
    let reader = std::thread::spawn(move || {
        let mut buf = Vec::new();
        out_pipe.read_to_end(&mut buf).ok();
        buf
    });
    let deadline = std::time::Instant::now() + std::time::Duration::from_secs(60);
    let status = loop {
        match child.try_wait() {
            Ok(Some(st)) => break Ok(st),
            Ok(None) if std::time::Instant::now() >= deadline => {
                let _ = child.kill();
                let _ = child.wait();
                return Ok(None);
            }
            Ok(None) => std::thread::sleep(std::time::Duration::from_millis(50)),
            Err(e) => break Err(e),
        }
    };
    let buf = reader.join().unwrap_or_default();
    match status {
        Ok(_) => Ok(Some(parse_ra_symbols(&String::from_utf8_lossy(&buf)))),
        Err(e) => Err(format!("rust-analyzer wait 失敗：{}", e)),
    }
}

/// graph.db nodes per-name counts (`graph_audit.py:154-161`): DB side kind
/// must include 'Test' (missing it flags every test fn as a false gap).
pub fn db_functions(conn: &Connection, path: &Path) -> HashMap<String, usize> {
    let resolved = crate::common::resolve(path);
    let Ok(mut stmt) = conn.prepare(
        "SELECT name, COUNT(*) FROM nodes WHERE file_path = ?1 \
         AND kind IN ('Function', 'Test') GROUP BY name",
    ) else {
        return HashMap::new();
    };
    let Ok(rows) = stmt.query_map([resolved.to_string_lossy().as_ref()], |r| {
        Ok((r.get::<_, String>(0)?, r.get::<_, i64>(1)?))
    }) else {
        return HashMap::new();
    };
    rows.filter_map(|r| r.ok())
        .map(|(n, c)| (n, c as usize))
        .collect()
}

#[derive(Debug, Clone)]
pub struct MissingItem {
    pub file: String,
    pub symbol: String,
    pub ra_count: usize,
    pub db_count: usize,
}

/// Test injection point (mirrors the Python `ra_lookup` parameter).
pub type RaLookup<'a> = &'a dyn Fn(&Path) -> Result<Option<OrderedCounter>, String>;

/// D1+D2 main flow (`graph_audit.py:164-213`): returns
/// `(risk, audited_count, missing, errors, total_ra, stderr_warns)`.
/// The per-file vacuous WARN lines go to the returned stderr buffer (the
/// lib never prints); the aggregate all-zero guard lives in the CLI.
#[allow(clippy::type_complexity)] // mirror of the frozen Python tuple shape
pub fn audit(
    repo: &Path,
    graph: &Path,
    all_files: bool,
    ra_lookup: Option<RaLookup>,
) -> Result<
    (
        Vec<(PathBuf, String, Vec<String>)>,
        usize,
        Vec<MissingItem>,
        Vec<String>,
        usize,
        Vec<String>,
    ),
    String,
> {
    let default_lookup = |p: &Path| ra_symbols(p);
    let lookup = ra_lookup.unwrap_or(&default_lookup);
    let files = scan_files(repo)?;
    let risk = risk_scan(&files);
    let scope: Vec<PathBuf> = if all_files {
        files
    } else {
        let mut s: Vec<PathBuf> = risk.iter().map(|(f, _, _)| f.clone()).collect();
        s.sort();
        s.dedup();
        s
    };
    let conn = connect_ro(graph)?;
    let mut missing = Vec::new();
    let mut errors = Vec::new();
    let mut warns = Vec::new();
    let mut total_ra: usize = 0;
    for f in &scope {
        let ra = match lookup(f) {
            Ok(Some(ra)) => ra,
            Ok(None) => {
                errors.push(format!("{}: rust-analyzer 逾時/失敗（跳過）", f.display()));
                continue;
            }
            Err(e) => return Err(e),
        };
        total_ra += ra.total();
        let db = db_functions(&conn, f);
        let size = std::fs::metadata(f).map(|m| m.len()).unwrap_or(0);
        if ra.is_empty() && size > 0 {
            warns.push(format!(
                "[WARN] rust-analyzer 對 {} 零輸出（格式 drift 或單檔 parse fail）——該檔對帳 vacuous，勿當乾淨讀\n",
                f.file_name().map(|n| n.to_string_lossy().into_owned()).unwrap_or_default()
            ));
        }
        for (name, ra_count) in ra.iter() {
            let db_count = db.get(name).copied().unwrap_or(0);
            if db_count < ra_count {
                missing.push(MissingItem {
                    file: f.display().to_string(),
                    symbol: name.to_string(),
                    ra_count,
                    db_count,
                });
            }
        }
    }
    Ok((risk, scope.len(), missing, errors, total_ra, warns))
}

/// The three env-gate failure messages, shared verbatim by the
/// `graph_audit` CLI and the `scip_refs --audit` wrapper (copying the
/// strings would drift on the first edit).
pub(crate) fn env_gate_messages() -> [&'static str; 3] {
    [
        "rust-analyzer 不在 PATH——rustup component add rust-analyzer",
        "graph.db 不存在：{graph}（完整度稽核需要它；新鮮度指標不保證存在）",
        "全部檔案 rust-analyzer 符號數為 0——輸出格式漂移或環境錯誤",
    ]
}

pub(crate) fn which(bin: &str) -> Option<PathBuf> {
    use std::os::unix::fs::PermissionsExt;
    let path = std::env::var_os("PATH")?;
    std::env::split_paths(&path)
        .map(|d| d.join(bin))
        .find(|p| {
            std::fs::metadata(p)
                .map(|m| m.is_file() && m.permissions().mode() & 0o111 != 0)
                .unwrap_or(false)
        })
}

/// Route a `code-reality graph_audit ...` invocation.
pub fn run(argv: &[&str]) -> ToolOutput {
    let Some((&_sub, toks)) = argv.split_first() else {
        return ToolOutput::fail("需提供子命令 graph_audit");
    };
    let values = match parse(&SPEC, toks) {
        Outcome::Help => {
            return ToolOutput {
                stdout: HELP.to_string(),
                stderr: String::new(),
                exit_code: 0,
            };
        }
        Outcome::Err(msg) => return ToolOutput::fail(msg),
        Outcome::Ok { values, .. } => values,
    };
    let repo = match required(&values, "--repo") {
        Ok(r) => PathBuf::from(r),
        Err(msg) => return ToolOutput::fail(msg),
    };
    let all_files = values.contains_key("--all");
    let as_json = values.contains_key("--json");
    let mut graph = values
        .get("--graph")
        .and_then(|v| v.clone())
        .map(PathBuf::from)
        .unwrap_or_else(|| graph_db_path(&repo));
    if graph.as_os_str().is_empty() {
        // Python Path("") == Path(".") — exists() is true there, and the
        // failure lands in connect_ro (exit 1), not the env gate (exit 2)
        graph = PathBuf::from(".");
    }

    let mut stderr = String::new();
    let gates = env_gate_messages();
    if which("rust-analyzer").is_none() {
        stderr.push_str(&crate::msg_line("FAIL", gates[0]));
        return ToolOutput { stdout: String::new(), stderr, exit_code: 2 };
    }
    if !graph.exists() {
        stderr.push_str(&crate::msg_line(
            "FAIL",
            &gates[1].replace("{graph}", &graph.display().to_string()),
        ));
        return ToolOutput { stdout: String::new(), stderr, exit_code: 2 };
    }

    let (risk, audited, missing, errors, total_ra, warns) = match audit(&repo, &graph, all_files, None)
    {
        Ok(v) => v,
        Err(e) => {
            // Python crashes uncaught here (exit 1, empty stdout)
            let mut out = ToolOutput::crash(e);
            out.stderr.insert_str(0, &stderr);
            return out;
        }
    };
    for w in warns {
        stderr.push_str(&w);
    }
    // aggregate vacuous guard: all-zero symbols across an audited scope
    if audited > 0 && total_ra == 0 {
        stderr.push_str(&crate::msg_line("FAIL", gates[2]));
        return ToolOutput { stdout: String::new(), stderr, exit_code: 2 };
    }

    let stdout = if as_json {
        let risk_files: Vec<serde_json::Value> = risk
            .iter()
            .map(|(f, t, o)| {
                serde_json::json!({
                    "file": f.display().to_string(),
                    "type": t,
                    "overlap": o,
                })
            })
            .collect();
        let missing_v: Vec<serde_json::Value> = missing
            .iter()
            .map(|m| {
                serde_json::json!({
                    "file": m.file,
                    "symbol": m.symbol,
                    "ra_count": m.ra_count,
                    "db_count": m.db_count,
                })
            })
            .collect();
        let errors_v: Vec<&String> = errors.iter().collect();
        // Python `print(json.dumps(...))` — the trailing newline is part of
        // the stdout byte face
        let mut body = crate::common::to_json_indent1(&serde_json::json!({
            "risk_files": risk_files,
            "audited_files": audited,
            "missing": missing_v,
            "errors": errors_v,
        }));
        body.push('\n');
        body
    } else {
        let mut out = String::new();
        out.push_str(&format!(
            "[OK] D1 風險掃描：{} 檔（同名方法 ≥2 impl 塊）\n",
            risk.len()
        ));
        out.push_str(&format!(
            "[OK] D2 對帳：{} 檔（rust-analyzer vs graph.db，{} 符號）\n",
            audited, total_ra
        ));
        for e in &errors {
            out.push_str(&format!("[WARN] {e}\n"));
        }
        if !missing.is_empty() {
            out.push_str(&format!(
                "[WARN] DB 缺差 {} 項（同鍵去重吃掉——head_matches_build 不反映此項）：\n",
                missing.len()
            ));
            let mut by_file: Vec<(String, Vec<&MissingItem>)> = Vec::new();
            for m in &missing {
                match by_file.iter_mut().find(|(f, _)| f == &m.file) {
                    Some((_, items)) => items.push(m),
                    None => by_file.push((m.file.clone(), vec![m])),
                }
            }
            by_file.sort_by(|a, b| a.0.cmp(&b.0));
            for (f, items) in &by_file {
                let syms: Vec<String> = items
                    .iter()
                    .map(|m| format!("{}({}/{})", m.symbol, m.db_count, m.ra_count))
                    .collect();
                out.push_str(&format!("  {f}: {}\n", syms.join(", ")));
            }
        } else {
            out.push_str("[OK] 無缺差\n");
        }
        out
    };
    let exit_code = if missing.is_empty() { 0 } else { 1 };
    ToolOutput { stdout, stderr, exit_code }
}
